Add 4337 Dependency section to the spec (#185)
alchemyplatform/ERC-6900-Ref-Implementation: standard/ERCs/erc-6900.md

- the account MUST clear all flags for the validation function, like `isGlobal`, `isSignatureValidation`, and `isUserOpValidation`.
- the account MUST remomve all hooks and SHOULD clear hook module states by calling `onUninstall` with the user-provided data for each hook, including both pre validation hooks and execution hooks, if required by user.

>   - the account MAY ignore the revert from `onUninstall` with try/catch depending on the design principle of the account.

Reviewer: Thoughts on this @erc6900/working-group ?

Reviewer 2: Seems fine. Adds some flexibility to the account.

Reviewer 3: 👍 It provides developers with flexibility when a clean state is not required in certain modules

This proposal includes several interfaces that build on ERC-4337. First, we standardize a set of modular functions that allow smart contract developers greater flexibility in bundling validation, execution, and hook logic. We also propose interfaces that provide methods for querying execution functions, validation functions, and hooks on a modular account. The rest of the interfaces describe a module's methods for exposing its modular functions and desired configuration, and the modular account's methods for installing and removing modules and allowing execution across modules and external addresses.

### ERC 4337 Dependency

ERC-6900's main objective is to create a secure and interoperable foundation through modular modules to increase the velocity and security of the Smart Account ecosystem and ultimately the wallet ecosystem.
Currently, the ERC-6900 standard enforces the ERC-4337 for the validation and execution, however, this does not dictate that ERC-6900 will continue to be tied to the ERC-4337.
It is likely that smart account builders will want to develop modular accounts that do not use ERC-4337 in the future, e.g., Native AA on rollups.

> Moreover, it is expected that the version of EntryPoint contract will continue to evolve until there is a protocol-level account abstraction.

Reviewer 2: +1 on those points. 6900 accounts can be used with or without 4337 support. It should be able to support different flavors of AA.

ERC-6900, having the objective to provide the secure foundation for modular smart account layer, does not try to be tied with a specific version of continuously evolving EntryPoint and AA architecture but rather position to function as a modular layer encompassing them.

To be in line with this, identical behavior and functions will exist for EntryPoint v6 with `UserOperation` and will continue to exist when a new EntryPoint version comes out with a different UserOp structure.

ERC-6900 could evolve with ERC-4337 becoming an extension through a separate ERC for ERC-6900 if more builders build on a new architecture, but in its current state, it enforces ERC-4337 within the standard, considering that the vast majority of smart accounts utilize ERC-4337.

> 

Reviewer: I'm uncertain whether we need this section until we actually begin to support a more flexible interface. Related issue: https://github.com/erc6900/resources/issues/14.

> To be in line with this, identical behavior and functions will exist for EntryPoint v6 with `UserOperation` and will continue to exist when a new EntryPoint version comes out with a different UserOp structure.

Is this line implying that `PackedUserOperation` as used in the current spec is interchangeable with ERC-4337 v0.6's `UserOperation`? I believe that there are some things stated in the spec that cannot be adequately supported without ERC-4337 v0.7, namely validation-associated execution hooks. 

Making the spec more flexible in this way probably requires further discussion.

Author: > I believe that there are some things stated in the spec that cannot be adequately supported without ERC-4337 v0.7, namely validation-associated execution hooks.

Could you elaborate further on validation-associated execution hooks and why it couldn't be supported on the `UserOperation` architecture?
Would love to learn further on that.

> I'm uncertain whether we need this section until we actually begin to support a more flexible interface. Related issue: https://github.com/erc6900/resources/issues/14.

Even if we don't enshrine 4337 or support a different interface/architecture yet, I think it's important that we state the objective of this ERC is on the modular account layer and not the 4337 bound account system. And that ERC 6900 views ERC 4337 more as a tool to enable it, and is flexible/open to new designs if new promising architecture comes up that builders are willing to build on.
I believe this will help this ERC be more future proof for changes in the future, and allow readers understand our focus.

Happy to hear what others think.

Reviewer 4: In my understanding validation associated execution hooks can optionally require the UserOperation context, which then use the `executeUserOp()` flow in order to also forward that context if needed. `executeUserOp()` is not present in v0.6. While I'm in favour of mentioning the potentially weakening 4337 dependency,  I'm a bit sceptical if the wording around inclusion of 4337v0.6, as 6900v0.8, as well as 7579 have been built from the ground up with 4337v0.7 in mind. Given the interface changes, to e.g. `validateUserOp()` supporting both versions would either require two plugin implementations or compatibility contracts, which would make things more complicated.

> 

Reviewer: We discussed this a bit with the working group yesterday. How do you feel about a trimmed down version of this @PowerStream3604 ?

```suggestion
### ERC-4337 Dependency

ERC-6900's main objective is to create a secure and interoperable foundation through modular accounts and modules to increase the velocity and security of the smart account ecosystem, and ultimately the wallet ecosystem. Currently, the standard prescribes ERC-4337 for one of its [modular account call flows](#overview). However, this does not dictate that ERC-6900 will continue to be tied to ERC-4337.

It is likely that smart account builders will want to develop modular accounts that do not use ERC-4337 in the future (e.g., native account abstraction on rollups). Moreover, it is expected that ERC-4337 and its interfaces and contracts will continue to evolve until there is a protocol-level account abstraction.

ERC-6900 will follow the research, development, and adoption in the space to address its objectives and create a secure foundation for modular accounts that may eventually be agnostic to the account abstraction mechanism used.
```

@nikita-quantstamp - would appreciate your help wordsmithing this!

Reviewer 5: I agree with @jaypaik's direction of not diving too deep into technical details in this section and mentioning of what we will keep vs what we will not keep in the future, as it is pretty hard to predict. Overall, I like it, I just added a few changes to tie it together. 

```suggestion
### ERC 4337 Dependency

ERC-6900's main objective is to create a secure and interoperable foundation through modular accounts and modules to increase the velocity and security of the smart account ecosystem, and ultimately the wallet ecosystem. Currently, the standard prescribes ERC-4337 for one of its [modular account call flows](#overview). However, this does not dictate that ERC-6900 will continue to be tied to ERC-4337.

It is likely that smart account builders will want to develop modular accounts that do not use ERC-4337 in the future (e.g., native account abstraction on rollups). Moreover, it is expected that ERC-4337 and its interfaces and contracts will continue to evolve until there is a protocol-level account abstraction.

In the current state of the AA ecosystem, it is tough to predict the direction the builders and industry will take, so ERC-6900 will follow the space's research, development, and adoption. The standard will do its best to address the objectives and create a secure foundation for modular accounts that may eventually be abstracted away from the infrastructure mechanism used.
```

Author: Appreciate the suggestion @jaypaik @nikita-quantstamp , what do you think of the below suggestion?
Tried to incorporate your suggestions while not losing the original flavor.

```suggestion
### ERC 4337 Dependency
	
ERC-6900's main objective is to create a secure and interoperable foundation through modular accounts and modules to increase the velocity and security of the smart account ecosystem, and ultimately the wallet ecosystem. Currently, the standard prescribes ERC-4337. e.g., [modular account call flows](#overview). However, this does not dictate that ERC-6900 will continue to be tied to ERC-4337.
	
It is likely that smart account builders will want to develop modular accounts that do not use ERC-4337 in the future (e.g., native account abstraction on rollups). Moreover, it is expected that ERC-4337 and its interfaces and contracts will continue to evolve until there is a protocol-level account abstraction.
	
The current ERC-6900 v0.8 does enforce EntryPoint v0.7. However, ERC-6900, having the objective to provide the secure foundation for modular smart account layer, does not try to be tied with a specific version of continuously evolving EntryPoint and AA architecture but rather position to function as a modular account layer flexibly utilizing/encompassing them as a tool.

In the current state of the AA ecosystem, it is tough to predict the direction the builders and industry will take, so ERC-6900 will evolve together with the space's research, development, and adoption. The standard will do its best to address the objectives and create a secure foundation for modular accounts that may eventually be abstracted away from the infrastructure mechanism used.

```

This proposal includes several interfaces that build on ERC-4337. First, we standardize a set of modular functions that allow smart contract developers greater flexibility in bundling validation, execution, and hook logic. We also propose interfaces that provide methods for querying execution functions, validation functions, and hooks on a modular account. The rest of the interfaces describe a module's methods for exposing its modular functions and desired configuration, and the modular account's methods for installing and removing modules and allowing execution across modules and external addresses.

> ### ERC 4337 Dependency

Reviewer: ```suggestion
### ERC-4337 Dependency
```

> ERC-6900's main objective is to create a secure and interoperable foundation through modular accounts and modules to increase the velocity and security of the smart account ecosystem, and ultimately the wallet ecosystem. Currently, the standard prescribes ERC-4337. e.g., [modular account call flows](#overview). However, this does not dictate that ERC-6900 will continue to be tied to ERC-4337.

Reviewer: ```suggestion
ERC-6900's main objective is to create a secure and interoperable foundation through modular accounts and modules to increase the velocity and security of the smart account ecosystem, and ultimately the wallet ecosystem. Currently, the standard prescribes ERC-4337 for one of its [modular account call flows](#overview). However, this does not dictate that ERC-6900 will continue to be tied to ERC-4337.
```
